### PR TITLE
BUG: fix handling of nan input for Time properties

### LINKED ISSUE #460: CADC SIA2 test failure
We started seeing this failure in the past week or so, seems to be more than transient now (but I haven't looked into the details yet, opening this issue to document the problem):

```
=================================== FAILURES ===================================
_____________________________ TestSIACadc.test_pos _____________________________

self = <[AttributeError("'Time' object has no attribute '_time'") raised in repr()] Time object at 0x7f5df1a91110>
val = array(nan), val2 = None, format = 'mjd', scale = None, precision = 3
in_subfmt = '*', out_subfmt = '*'

    def _get_time_fmt(self, val, val2, format, scale, precision, in_subfmt, out_subfmt):
        """
        Given the supplied val, val2, format and scale try to instantiate
        the corresponding TimeFormat class to convert the input values into
        the internal jd1 and jd2.
    
        If format is `None` and the input is a string-type or object array then
        guess available formats and stop when one matches.
        """
        if format is None and (
            val.dtype.kind in ("S", "U", "O", "M") or val.dtype.names
        ):
            # Input is a string, object, datetime, or a table-like ndarray
            # (structured array, recarray). These input types can be
            # uniquely identified by the format classes.
            formats = [
                (name, cls)
                for name, cls in self.FORMATS.items()
                if issubclass(cls, TimeUnique)
            ]
    
            # AstropyTime is a pseudo-format that isn't in the TIME_FORMATS registry,
            # but try to guess it at the end.
            formats.append(("astropy_time", TimeAstropyTime))
    
        elif not (isinstance(format, str) and format.lower() in self.FORMATS):
            if format is None:
                raise ValueError(
                    "No time format was given, and the input is not unique"
                )
            else:
                raise ValueError(
                    f"Format {format!r} is not one of the allowed formats "
                    f"{sorted(self.FORMATS)}"
                )
        else:
            formats = [(format, self.FORMATS[format])]
    
        assert formats
        problems = {}
        for name, cls in formats:
            try:
>               return cls(val, val2, scale, precision, in_subfmt, out_subfmt)

.tox/py311-test-alldeps-online/lib/python3.11/site-packages/astropy/time/core.py:597: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.tox/py311-test-alldeps-online/lib/python3.11/site-packages/astropy/time/formats.py:159: in __init__
    val1, val2 = self._check_val_type(val1, val2)
.tox/py311-test-alldeps-online/lib/python3.11/site-packages/astropy/time/formats.py:483: in _check_val_type
    val1, val2 = super()._check_val_type(val1, val2)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <astropy.time.formats.TimeMJD object at 0x7f5df1e0f610>
val1 = array(nan), val2 = None

    def _check_val_type(self, val1, val2):
        """Input value validation, typically overridden by derived classes."""
        # val1 cannot contain nan, but val2 can contain nan
        isfinite1 = np.isfinite(val1)
        if val1.size > 1:  # Calling .all() on a scalar is surprisingly slow
            isfinite1 = (
                isfinite1.all()
            )  # Note: arr.all() about 3x faster than np.all(arr)
        elif val1.size == 0:
            isfinite1 = False
        ok1 = (
            val1.dtype.kind == "f"

    def _get_time_fmt(self, val, val2, format, scale, precision, in_subfmt, out_subfmt):
        """
        Given the supplied val, val2, format and scale try to instantiate
        the corresponding TimeFormat class to convert the input values into
        the internal jd1 and jd2.
    
        If format is `None` and the input is a string-type or object array then
        guess available formats and stop when one matches.
        """
        if format is None and (
            val.dtype.kind in ("S", "U", "O", "M") or val.dtype.names
        ):
            # Input is a string, object, datetime, or a table-like ndarray
            # (structured array, recarray). These input types can be
            # uniquely identified by the format classes.
            formats = [
                (name, cls)
                for name, cls in self.FORMATS.items()
                if issubclass(cls, TimeUnique)
            ]
    
            # AstropyTime is a pseudo-format that isn't in the TIME_FORMATS registry,
            # but try to guess it at the end.
            formats.append(("astropy_time", TimeAstropyTime))
    
        elif not (isinstance(format, str) and format.lower() in self.FORMATS):
            if format is None:
                raise ValueError(
                    "No time format was given, and the input is not unique"
                )
            else:
                raise ValueError(
                    f"Format {format!r} is not one of the allowed formats "
                    f"{sorted(self.FORMATS)}"
                )
        else:
            formats = [(format, self.FORMATS[format])]
    
        assert formats
        problems = {}
        for name, cls in formats:
            try:
                return cls(val, val2, scale, precision, in_subfmt, out_subfmt)
            except UnitConversionError:
                raise
            except (ValueError, TypeError) as err:
                # If ``format`` specified then there is only one possibility, so raise
                # immediately and include the upstream exception message to make it
                # easier for user to see what is wrong.
                if len(formats) == 1:
>                   raise ValueError(
                        f"Input values did not match the format class {format}:"
                        + os.linesep
                        + f"{err.__class__.__name__}: {err}"
                    ) from err
E                   ValueError: Input values did not match the format class mjd:
E                   TypeError: Input values for mjd class must be finite doubles

.tox/py311-test-alldeps-online/lib/python3.11/site-packages/astropy/time/core.py:605: ValueError
=========================== short test summary info ============================
XFAIL pyvo/io/vosi/tests/test_tables.py::TestTables::test_wrong_flag
FAILED pyvo/dal/tests/test_sia2_remote.py::TestSIACadc::test_pos - ValueError...
=== 1 failed, 318 passed, 1 xfailed, 1 xpassed, 1 rerun in 176.94s (0:02:56) ===
```

cc @andamian

### PR DESCRIPTION
**Pull request by bsipocz**
This should fix the current test failure with the IRAS record from CADC. 

Fix #460

### CODE REVIEW

**Comment by bsipocz** [comment]
(ignore the changelog failure, I was testing the milestone check and while it's been fixed, it doesn't trigger a rerun and thus a green checkmark)

File: pyvo/dal/sia2.py [hunk 1]
```diff
@@ -884,7 +885,11 @@ def t_min(self):
         products result of the combination of multiple frames, min time must
         be the minimum of the start times
         """
-        return Time(self.get('t_min'), format='mjd')
+        t_min = self.get('t_min')
+        if np.isfinite(t_min):
+            return Time(t_min, format='mjd')
+        else:
```
Comment (andamian) [inline pyvo/dal/sia2.py:891]: Wondering whether we should leak the `nan` here or just return `None`. @msdemlei does `nan` provide more meaning in this context?

Comment (bsipocz) [reply pyvo/dal/sia2.py:891]: we already leak nan-s for many of the other elements, exptimes, etc. (It's only `Time` that chokes on them, Quantity handles it well, etc.)

Comment (bsipocz) [reply pyvo/dal/sia2.py:891]: see the copy-pasted examples in https://github.com/astropy/pyvo/issues/460#issuecomment-1650572142

Comment (andamian) [reply pyvo/dal/sia2.py:891]: OK. I've looked at that but I thought it might be from some representation inside `pyvo`

**Review by tomdonaldson — APPROVED** [review]
Thanks @bsipocz .  This is clearly better than before.

I do agree that it feels like there would be a better way to deal with this.  It's unclear to me how (or currently if) this property is used, so maybe that would help, but I wonder how useful or important for this getter to return a `Time` object.  Any opportunity for a getter to fail makes it more fragile.  

With this one now returning either a time or something that isn't `np.finite()`, a consumer of the value likely still needs to check which it is, so they may as well be the ones to create the `Time` object if that's what they want.

**Comment by bsipocz** [comment]
The issue with returning a Time object is that there isn't really an empty time object, as opposed to e.g. a Quantity/etc.

That being said, I'm happy to return whatever you think is better. And I have just checked and SIA we return None for the `dateobs` property, so that's an easy thing to do here, too.

**Comment by tomdonaldson** [comment]
> The issue with returning a Time object is that there isn't really an empty time object, as opposed to e.g. a Quantity/etc.
> 
> That being said, I'm happy to return whatever you think is better. And I have just checked and SIA we return None for the `dateobs` property, so that's an easy thing to do here, too.

Yes, maybe None is a little better in that it might be a more intuitive thing for a consumer to check on, though they still have to check, so it circles back to me not knowing how this is intended to be used.  I'm happy either way for this PR.

**Comment by bsipocz** [comment]
> so it circles back to me not knowing how this is intended to be used

honestly, I don't understand why pyvo is so heavy peppered through with these properties, IMO it that makes the API more complex, but it's likely I miss some point somewhere.

**Comment by msdemlei** [comment]
On Wed, Jul 26, 2023 at 10:57:09AM -0700, Adrian wrote:
> Wondering whether we should leak the `nan` here or just return
> `None`. @msdemlei does `nan` provide more meaning in this context?

VOTable, IMHO regrettably, does not distinguish None and NaN.
In TABLEDATA and BINARY2 serialisation, we could (by and large)
maintain this distinction, but the use of NaN as a NULL value is far
too widespread to bother.

So, nothing that's not already broken will break if you identify None
and NaN for anything coming out of a VOTable.

**Comment by msdemlei** [comment]
On Wed, Jul 26, 2023 at 03:15:27PM -0700, Brigitta Sipőcz wrote:
> honestly, I don't understand why pyvo is so heavy peppered through
> with these properties, IMO it that makes the API more complex, but
> it's likely I miss some point somewhere.

Tangentially: I'm not a fan of the wide use of properties either.
Not enough to break existing APIs, but I think for new APIs we should
encourage less magic of this kind, even if it means that we have
different API flavours in one package.

**Comment by bsipocz** [comment]
OK, so everyone seems to be happy with the None case, so I go ahead and merge this. Thanks for the quick review cycle!

**Pull request merged**